### ISSUE TITLE
mustRunAfter not respected when running tasks in parallel

### ISSUE THREAD

**Issue by rkrisztian** [gradle/gradle#10549]
I'm using Gradle 5.6.1 on RHEL7 (Linux), and encountered a weird problem on a real project, which I reproduced on a very simplified test project: [test-parallel.zip](https://github.com/gradle/gradle/files/3596769/test-parallel.zip).

This test project has 3 subprojects, all doing the same: making a bear find a bee hive, steal it, check if the honey in it tastes good, and then throw away the hive. (Excuse me for not having a more meaningful example.)

The reason I have this many tasks is that in the real project I use the [gradle-docker-plugin](https://github.com/bmuschko/gradle-docker-plugin), which makes me create a task for each docker operation. And I wanted those tasks to be flexible, so they depend on each other, but some tasks can be issued separately too (depending on command-line flags), so that's why you see {{mustRunAfter}} in places where a {{dependOn}} association would be just fine too.

Now, all tasks run fine when executed sequentially:

```none
$ gw bearCheckIfHoneyTastesGood --continue --no-parallel

> Task :test1:bearFindBeeHive
> Task :test1:bearStealBeeHive
> Task :test1:bearInspectBeeHive
> Task :test1:bearRequireBeeHive
> Task :test1:bearCheckIfHoneyTastesGood
> Task :test1:bearThrowAwayBeeHive
> Task :test2:bearFindBeeHive
> Task :test2:bearStealBeeHive
> Task :test2:bearInspectBeeHive
> Task :test2:bearRequireBeeHive
> Task :test2:bearCheckIfHoneyTastesGood
> Task :test2:bearThrowAwayBeeHive
> Task :test3:bearFindBeeHive
> Task :test3:bearStealBeeHive
> Task :test3:bearInspectBeeHive
> Task :test3:bearRequireBeeHive
> Task :test3:bearCheckIfHoneyTastesGood
> Task :test3:bearThrowAwayBeeHive
```

When I run them in parallel, I expect that there is only one bee hive stolen at a time (in the real project the docker containers would conflict on resources). However, I see build failures occurring with a random probability, like:

* multiple bee hives stolen at the same time (`mustRunAfter` not respected across projects)
    * sometimes without this detected in the Gradle code (I assume that is my fault for not using a thread-safe data type, or is it?)
* Task `bearRequireBeeHive` executed before `bearStealBeeHive` within the same project (`mustRunAfter` not respected within the same project), which is even weirder.

```
$ gw bearCheckIfHoneyTastesGood --continue

> Task :test2:bearFindBeeHive
> Task :test1:bearFindBeeHive
> Task :test3:bearFindBeeHive
> Task :test2:bearStealBeeHive
> Task :test1:bearStealBeeHive FAILED
> Task :test2:bearRequireBeeHive FAILED
> Task :test1:bearInspectBeeHive
> Task :test1:bearRequireBeeHive
> Task :test1:bearThrowAwayBeeHive
> Task :test2:bearInspectBeeHive
> Task :test2:bearThrowAwayBeeHive
> Task :test3:bearStealBeeHive
> Task :test3:bearInspectBeeHive
> Task :test3:bearRequireBeeHive
> Task :test3:bearCheckIfHoneyTastesGood
> Task :test3:bearThrowAwayBeeHive

FAILURE: Build completed with 2 failures.

1: Task failed with an exception.
-----------
* Where:
Build file '/home/rkrisztian/projects/testProjects/test-parallel/test1/build.gradle' line: 16

* What went wrong:
Execution failed for task ':test1:bearStealBeeHive'.
> Only one bee hive should be stolen at a time.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
==============================================================================

2: Task failed with an exception.
-----------
* Where:
Build file '/home/rkrisztian/projects/testProjects/test-parallel/test2/build.gradle' line: 34

* What went wrong:
Execution failed for task ':test2:bearRequireBeeHive'.
> Bee hive not OK

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
==============================================================================

* Get more help at https://help.gradle.org

BUILD FAILED in 516ms
```

I need your help with this: is this a Gradle bug, or am I doing anything wrong?

Thanks in advance.

**Comment by wolfs** [comment]
It seems to me that what happens here is the following:
There is only a `finalizedBy` relationship between `bearCheckIfHoneyTastesGood` and `bearThrowAwayBeeHive`, so Gradle doesn't know up-front that `bearThrowAwayBeeHive` needs to run, since it is not part of the requested tasks.

And since it isn't part of the required tasks, e.g. `:test1:throwAwayBeeHive` doesn't block `:test2:bearStealBeeHive` from running.

In theory, Gradle knows that the finalizer will run at some point if one of the requested tasks executed or will be executed. So we could make this use-case work.

Do you want more something like shared resources which can only be used by one task at a time? In Gradle 6.0, we introduce a concept for exactly this: https://github.com/gradle/gradle/pull/9914

Maybe this is better suited for your needs.

**Comment by rkrisztian** [comment]
Thanks for the explanation.

In theory, yes, **the shared resources concept would help** for this problem, **if we can make sure that the same resource can be (atomically) reserved by all tasks** executed between `bearStealBeeHive` and `bearThrowAwayBeeHive`, because the first task allocates the resources, and the last task frees it up in our case. So if there is still any possibility of race conditions between project `test1` and `test2`, then our problem is still not solved, and it is not clear at a first glance that the PR can handle this too. I read the bug report that originated the PR though and it was very similar to my problem.

However, I would like to know  **if the phenomenon that lead to task `bearRequireBeeHive` executed before `bearInspectBeeHive` is really the same** and can be solved the same way, as the difference there is that those tasks were executed within the same project, so it is not obvious to me if the shared resources concept would help that case too.

In general, **is there a documentation on `mustRunAfter` that it may not work for finalizer tasks**? I tried to look up this information, and I found none, which lead me to this "bug" in the first place. So if there is none, I would at least ask for a little documentation update. Because someone else might run into the same problem in the future again, for something different, which may or may not be solved with a shared resource configuration.

**Comment by wolfs** [comment]
There is some workaround you can use: you can add a `mustRunAfter` between the task which is finalized and the task which should run after it as well. That should give you the desired effect. So in your case, `:test2:bearStealBeeHive.mustRunAfter(':test1:bearCheckIfHoneyTastesGood')` should do the trick.

> In theory, yes, **the shared resources concept would help** for this problem, **if we can make sure that the same resource can be (atomically) reserved by all tasks** executed between `bearStealBeeHive` and `bearThrowAwayBeeHive`, because the first task allocates the resources, and the last task frees it up in our case.

IIUC, shared resources currently wouldn't solve this problems currently. Though We may implement exactly what you describe above in the future.

**Comment by rkrisztian** [comment]
I tried your suggestion, but it seems that no matter what I do, my `mustRunAfter` requests are still ignored. And I'm also failing to see how the suggestion results in correctness. I need to make sure that only one project executes any task between `bearStealBeeHive` and `bearThrowAwayBeeHive` (inclusively). 

That's what I've been trying to say regarding "atomic" resource allocation for a task graph, i.e. that a group of tasks collectively want to use a shared resource (with the first task allocating it, last task freeing it up), and not to be treated as a list of individual tasks each one expected to let go of the resource after executed, which is not true in my case.

**Comment by wolfs** [comment]
The following works for me:
In the top-level `build.gradle`, I ensure that the first task of the chain and finalizer of the first task of the chain runs after the last task of the chain and the finalizer of the last task of the chain:
```
private setTaskOrderBetweenProjects(projectBefore, projectAfter) {
    tasks.getByPath("$projectAfter:bearStealBeeHive").mustRunAfter "$projectBefore:bearThrowAwayBeeHive"
    tasks.getByPath("$projectAfter:bearStealBeeHive").mustRunAfter "$projectBefore:bearCheckIfHoneyTastesGood"
    tasks.getByPath("$projectAfter:bearRequireBeeHive").mustRunAfter "$projectBefore:bearThrowAwayBeeHive"
    tasks.getByPath("$projectAfter:bearRequireBeeHive").mustRunAfter "$projectBefore:bearCheckIfHoneyTastesGood"
}
```

Moreover, in the individual test projects, I make sure that `bearRequireBeeHive` runs after `bearStealBeeHive` as well, which is finalized of `bearInspectBeeHive`:

```
task('bearRequireBeeHive') {
      mustRunAfter bearInspectBeeHive
      mustRunAfter bearStealBeeHive

...
}
```

After these changes, the tasks execute as expected.

**Comment by rkrisztian** [comment]
Thank you so much, the same **worked** in my real project too!

Although we could at least find a workaround, having to configure this many associations is a bit **less than ideal**, to be honest. One would imagine that the following would normally be enough:

```groovy
    tasks.getByPath("$projectAfter:bearStealBeeHive").mustRunAfter "$projectBefore:bearThrowAwayBeeHive"
```

(I didn't add the `bearRequireBeeHive` one, because in my real project, we would run only the full chain in parallel. Partial chain executions are only considered for individual subprojects due to resource conflicts.)

Therefore, I would say, please consider at least documenting this limitation of `mustRunAfter`, if not adding full support for task graphs with finalizers, as [said above](https://github.com/gradle/gradle/issues/10549#issuecomment-531471299). It is no longer urgent for me, knowing that there is a fine workaround.

**Comment by rkrisztian** [comment]
Sadly the given workaround may break once `tasks.getByPath` gets removed from Gradle due to the pushing of the [task configuration avoidance API](https://docs.gradle.org/current/userguide/task_configuration_avoidance.html). Plus my memory is getting vague here, but I've read somewhere that it's not a good design to call such a method because we are digging into the internals of other projects.

**Comment by erdi** [comment]
FWIW, I've just hit this issue. 

It would be grand to get this fixed because it's actually quite surprising to have `mustRunAfter` ignored when the argument is a task that is a finaliser and for it to only happen when running in parallel mode so you end up scratching your head and doubting your build configuration... In the meantime I will use the workaround of also adding `mustRunAfter` for the finalized task as well, thanks for providing that suggestion @wolfs.

**Comment by wolfs** [comment]
I tried with Gradle 7.6 and this seems to work now: https://ge.gradle.org/s/zgn35dykqcus2/timeline


![Image](https://user-images.githubusercontent.com/423186/212037964-9efd8da2-c9e9-4a0e-977c-6ac1ebca0d67.png)
